### DISCUSSION TITLE
Typing a function that receives page and params

### DISCUSSION THREAD

**Discussion by jinkook0126** [react-navigation/react-navigation#11982] [Q&A]
Hello, I'm creating a function to be used in the Navbar. This function takes page and params as arguments and navigates to the desired page. I worked on it with ChatGPT, but there's a part that's not working properly, so I'm asking for your opinion.

```
export type RootStackParamList = {
  Home: NavigatorScreenParams<HomeTabParamList>;
  Profile: { id: string };
  Feed: { postId: number } | undefined;
  Faq: undefined;
};
```

Given the RootStackParamList type above, I created a function as follows:

```
const navigateToRoute = <T extends keyof RootStackParamList>(
  routeName: T,
  params?: RootStackParamList[T],
) => {
  if (params !== undefined) {
    navigation.navigate<T>(routeName, params);
  } else {
    navigation.navigate<T>(routeName);
  }
};
```

However, I'm getting an error that the routeName and params in the navigation.navigate() part do not match the types. It seems that this is happening because the generic type is not being inferred correctly. How can I fix this?

Below is the error.

```
Argument of type '[T, RootStackParamList[T] & {}]' is not assignable to parameter of type 'T extends unknown ? undefined extends RootParamList[T] ? [screen: T] | [screen: T, params: RootParamList[T]] : [screen: T, params: RootParamList[T]] : never'.ts
```

I haven't been able to work this out for days now, get me out of hell

**Comment by satya164** [comment]
Your logic is not correct. The function you wrote has `params` as optional, but whether params can be optional or not depends on the route name - it can't always be optional.

Not sure why you're trying to implement a function that is the same as `navigation.navigate`, but if you really want to do that, it's best to copy the [function signature](https://github.com/react-navigation/react-navigation/blob/11e83b38056bacdfddddd542e980b90fb5fa6222/packages/core/src/types.tsx#L238C11-L251C4) instead of trying to implement your own which may have different semantics.

```ts
const navigateToRoute = <RouteName extends keyof RootStackParamList>(
  ...args: RouteName extends unknown
    ? undefined extends RootStackParamList[RouteName]
      ?
          | [screen: RouteName]
          | [screen: RouteName, params: RootStackParamList[RouteName]]
      : [screen: RouteName, params: RootStackParamList[RouteName]]
    : never
) => {
  navigation.navigate<RouteName>(...args);
};
```

**Reply by jinkook0126** [reply]
Thank you for your previous response.

I'm not trying to implement navigation.navigate directly. Instead, I want to create a single function that handles routing based on login status and level for all 10 menus. Each button should call this function and pass the desired page as a parameter.

For example:
```
const onNavigate = (page, params) => {
  // ...some logic...
  navigation.navigate(page, params);
};

```
The problem is that page and params don't have type inference, so I tried to create a similar logic or wanted to share a solution.
